src/query/query_visitor.rs

```diff
@@ -0,0 +1,225 @@
+//!     }
+//! }
+//!
+//! impl<'ast> QueryVisitor<'ast> for FieldsCounter {
+//!     fn visit_field(&mut self, node: &'ast Field) {
+//!         self.count += 1
+//!     }
+//! }
+//!
+//! fn main() {
+//!     let mut number_of_type = FieldsCounter::new();
+//!
+//!     let doc = parse_query(r#"
+//!         query TestQuery {
+//!             users {
+//!                 id
+//!                 country {
+//!                     id
+//!                 }
+//!             }
+//!         }
+//!     "#).expect("Failed to parse query");
+//!
+//!     walk_document(&mut number_of_type, &doc);
+//!
+//!     assert_eq!(number_of_type.count, 2);
+//! }
+//! ```
+//!
+//! [`QueryVisitor`]: /graphql_parser/query/query_visitor/trait.QueryVisitor.html
+//! [`walk_document`]: /graphql_parser/query/query_visitor/fn.walk_document.html
+
+#![allow(unused_variables)]
+
+use super::ast::*;
+
+/// Trait for easy query syntax tree traversal.
+///
+/// See [module docs](/graphql_parser/query/query_visitor/index.html) for more info.
+pub trait QueryVisitor<'ast> {
+    fn visit_document(&mut self, node: &'ast Document) {}
+
+    fn visit_definition(&mut self, node: &'ast Definition) {}
+
+    fn visit_fragment_definition(&mut self, node: &'ast FragmentDefinition) {}
+
+    fn visit_operation_definition(&mut self, node: &'ast OperationDefinition) {}
+
+    fn visit_query(&mut self, node: &'ast Query) {}
+
+    fn visit_mutation(&mut self, node: &'ast Mutation) {}
+
+    fn visit_subscription(&mut self, node: &'ast Subscription) {}
+
+    fn visit_selection_set(&mut self, node: &'ast SelectionSet) {}
+
+    fn visit_variable_definition(&mut self, node: &'ast VariableDefinition) {}
+
+    fn visit_selection(&mut self, node: &'ast Selection) {}
+
+    fn visit_field(&mut self, node: &'ast Field) {}
+
+    fn visit_fragment_spread(&mut self, node: &'ast FragmentSpread) {}
+
+    fn visit_inline_fragment(&mut self, node: &'ast InlineFragment) {}
+}
+
+
+/// Walk a query syntax tree and call the visitor methods for each type of node.
+///
+/// This function is how you should initiate a visitor.
+pub fn walk_document<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Document) {
+    visitor.visit_document(node);
+    for def in &node.definitions {
+        walk_definition(visitor, def);
+    }
+}
+
+fn walk_definition<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast Definition) {
+    use super::ast::Definition::*;
+
+    visitor.visit_definition(node);
+    match node {
+        Operation(inner) => {
+            visitor.visit_operation_definition(inner);
+            walk_operation_definition(visitor, inner);
+        },
+        Fragment(inner) => {
+            visitor.visit_fragment_definition(inner);
+            walk_fragment_definition(visitor, inner);
+        },
+    }
+}
+
+fn walk_fragment_definition<'ast, V: QueryVisitor<'ast>>(visitor: &mut V, node: &'ast FragmentDefinition) {
+    visitor.visit_fragment_definition(node);
+}
```

Comment:
Would it make sense to further walk down the fragment fields?

Reply:
Addressed in db9310e 👍